src/shared/pkcs11-util.c

```diff
@@ -896,13 +897,11 @@ int pkcs11_token_find_private_key(
 
                 switch (attributes[i].type) {
                 case CKA_CLASS: {
-                        CK_OBJECT_CLASS c;
-
-                        if (attributes[i].ulValueLen != sizeof(c))
+                        if (attributes[i].ulValueLen != sizeof(CK_OBJECT_CLASS))
                                 return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Invalid PKCS#11 CKA_CLASS attribute size.");
 
-                        memcpy(&c, attributes[i].pValue, sizeof(c));
-                        if (c != CKO_PRIVATE_KEY)
+                        CK_OBJECT_CLASS *class = (CK_OBJECT_CLASS*) attributes[i].pValue;
```

Comment:
this is only safe if we can assume alignment. can we?

Reply:
PKCS#11 2.40:
```
typedef unsigned long int CK_ULONG;
typedef CK_ULONG CK_OBJECT_CLASS;
```
PKCS#11 3.0:
```
typedef unsigned long int CK_ULONG;
typedef CK_ULONG CK_OBJECT_CLASS;
```
PKCS#11 3.1:
```
typedef unsigned long int CK_ULONG;
typedef CK_ULONG CK_OBJECT_CLASS;
```
I think it is safe to use.

Reply:
if you want to make me happy you add:

```c
assert_cc(alignof(*attributes[i].pValue) == alignof(CK_OBJECT_CLASS));
```
with that in place the assumption is put in code, and if it's not true anymore we'll fail compilation

Reply:
Sorry, I understood you wrong here. The PKCS#11 2.40 and above guarantees a proper alignment for *pValue (see description of CK_ATTRIBUTE [here](http://docs.oasis-open.org/pkcs11/pkcs11-base/v2.40/os/pkcs11-base-v2.40-os.html#_Toc416959689)):

> Note that pValue is a “void” pointer, facilitating the passing of arbitrary values.  Both the application and Cryptoki library MUST ensure that the pointer can be safely cast to the expected type (i.e., without word-alignment errors).

I think we can consider p11-kit as "the application" and be sure it guarantees proper alignment also.

Reply:
Will [this comment](https://github.com/p11-glue/p11-kit/issues/626#issuecomment-2050766505) make you happy, or should we wait until the issue is closed?

Reply:
The current code won't work with wrong alignment anyway, because C_FindObjectsInit accepts only aligned CK_ATTRIBUTEs, so I mark as resolved.